File: src/lae/core/taskspaces.py
"""Taskspace visibility helpers for Waybar and navigation."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from lae.core.models import SessionState

DEFAULT_MIN_VISIBLE_WORKSPACES = 5


def visible_default_workspace_count(
    state: SessionState,
    allowed: list[str],
    active_rel: int,
    occupied: set[int],
) -> int:
    """How many workspace slots to show in Waybar for the current taskspace."""
    from lae.core.models import ContextMode

    total = len(allowed)
    if state.context_mode == ContextMode.task:
        return total

    highest_occupied = max(occupied) if occupied else 0
    visible = max(active_rel, highest_occupied, DEFAULT_MIN_VISIBLE_WORKSPACES)
    return min(visible, total, 10)

File: crates/lae-core/src/taskspaces.rs
use std::collections::HashSet;

use crate::models::{ContextMode, SessionState};

pub const DEFAULT_MIN_VISIBLE_WORKSPACES: u32 = 5;

pub fn visible_default_workspace_count(
    state: &SessionState,
    allowed: &[String],
    active_rel: i32,
    occupied: &HashSet<i32>,
) -> u32 {
    let total = allowed.len() as u32;
    if state.context_mode == ContextMode::Task {
        return total;
    }

    // Default taskspace: show at least DEFAULT_MIN_VISIBLE_WORKSPACES slots.
    let highest_occupied = occupied.iter().copied().max().unwrap_or(0) as u32;
    let active = active_rel.max(0) as u32;
    let visible = active
        .max(highest_occupied)
        .max(DEFAULT_MIN_VISIBLE_WORKSPACES);
    visible.min(total).min(10)
}
